rust/toolchain.bzl

```diff
@@ -301,3 +301,170 @@ See @rules_rust//rust:repositories.bzl for examples of defining the @rust_cpuX r
 with the actual binaries and libraries.
 """,
 )
+
+def _cargo_tool_impl(ctx):
+    toolchain = ctx.toolchains[str(Label("//rust:toolchain"))]
+
+    # Executables must be produced within the rule marked executable. To
+    # satisfy this, a symlink is created.
+    cargo = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.symlink(
+        output = cargo,
+        target_file = toolchain.cargo,
+        is_executable = True,
+    )
+
+    return [DefaultInfo(
+        files = depset([toolchain.cargo]),
+        executable = cargo,
+    )]
+
+_cargo_tool = rule(
```

Comment:
Regarding naming, I'm wondering if it makes sense to encode some/all of these things in the name:

* that this is the executable and all its runtime dependencies (data dependencies)
  * we used `_files` suffix for that in C++ rules, but that makes the fact that there is an executable a little bit less clear
* that this tool has been taken from the current Rust toolchain
  * maybe put `current` into the name?
* that this tool has been taken from the current Rust toolchain in the exec configuration (so it's supposed to be runnable on the execution platform, aka my computer, not on the computer where my Rust library I'm building using rules rust runs on)
  * maybe put "exec" in the name?
  
By maybe I really mean let's discuss because maybe there are better ways to communicate these things to the user and/or we shouldn't communicate them at all?

Reply:
I like `current_exec_{tool}` but I'm not sure about `_files`. I don't think we should guarantee that each target has everything it needs to be fully functional. I don't think the `cargo` tool should come with `rustc` or `rustc` come with `rustc_lib`. If a user is interested in wiring all that up, then I'd want to ensure all the components they're looking for are there, but the tools are really more for edge case stuff where you need a particular binary or source file for some reason. Which would mean being able to ask for `ctx.executable.rustc` in a rule would be the nicest pattern IMO.

Reply:
Ok, I've done a complete 180 here and added runfiles so the tools were runnable. This made it actually testable and still maintains some ability to parse just the file/binary requested.

Reply:
I was thinking of opening an actual discussion for naming but it feels a bit over kill. I've made the rule private so the only questions now are, do we like the idea of these targets (PR description has been updated) and how should they be named if we like them.

cc @illicitonion @dfreese

Reply:
For the record, I did end up going with `current_exec_{tool}_files` since the targets may have runfiles depending on which one is requested.

Reply:
I don't have any strong opinions here :)